pkg/storage/tsdb/tombstones.go

```diff
@@ -0,0 +1,360 @@
+	// the progress
+	tombstoneExists, err := m.TombstoneExists(ctx, tombstone.RequestID, tombstone.State)
+	if err != nil {
+		level.Error(m.logger).Log("msg", "unable to check if the same tombstone already exists", "requestID", tombstone.RequestID, "err", err)
+	} else if tombstoneExists {
+		return ErrTombstoneAlreadyExists
+	}
+
+	return errors.Wrap(m.bkt.Upload(ctx, fullTombstonePath, bytes.NewReader(data)), "upload tombstone file")
+}
+
+func (m *TombstoneManager) TombstoneExists(ctx context.Context, requestID string, state BlockDeleteRequestState) (bool, error) {
+	fullTombstonePath := path.Join(TombstonePath, getTombstoneFileName(requestID, state))
+	exists, err := m.bkt.Exists(ctx, fullTombstonePath)
+
+	if exists || err != nil {
+		return exists, err
+	}
+
+	return false, nil
+}
+
+func (m *TombstoneManager) GetDeleteRequestByIDForUser(ctx context.Context, requestID string) (*Tombstone, error) {
+	found := []*Tombstone{}
+
+	for _, state := range AllDeletionStates {
+		filename := getTombstoneFileName(requestID, state)
+		exists, err := m.TombstoneExists(ctx, requestID, state)
+		if err != nil {
+			return nil, err
+		}
+
+		if exists {
+			t, err := m.ReadTombstoneFile(ctx, path.Join(TombstonePath, filename))
+			if err != nil {
+				return nil, err
+			}
+			found = append(found, t)
+		}
+	}
+
+	if len(found) == 0 {
+		return nil, nil
+	}
+
+	// If there are multiple tombstones with the same request id but different state, want to return only the latest one
+	// The older states will be cleaned up by the compactor
+	return found[len(found)-1], nil
+
+}
+
+func (m *TombstoneManager) GetAllDeleteRequestsForUser(ctx context.Context) ([]*Tombstone, error) {
+	// add all the tombstones to a map and check for duplicates,
+	// if a key exists with the same request ID (but two different states)
+	tombstoneMap := make(map[string]*Tombstone)
+	err := m.bkt.Iter(ctx, TombstonePath, func(s string) error {
+		t, err := m.ReadTombstoneFile(ctx, s)
+		if err != nil {
+			return err
+		}
+
+		if _, exists := tombstoneMap[t.RequestID]; !exists {
+			tombstoneMap[t.RequestID] = t
+		} else {
+			// if there is more than one tombstone for a given request,
+			// we only want to return the latest state. The older file
+			// will be cleaned by the compactor
+			newT, err := m.getLatestTombstateByState(t, tombstoneMap[t.RequestID])
+			if err != nil {
+				return err
+			}
+			tombstoneMap[t.RequestID] = newT
+		}
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	deletionRequests := []*Tombstone{}
+	for _, t := range tombstoneMap {
+		deletionRequests = append(deletionRequests, t)
+	}
+
+	return deletionRequests, nil
+}
+
+func (m *TombstoneManager) getLatestTombstateByState(a *Tombstone, b *Tombstone) (*Tombstone, error) {
+	orderA, err := a.State.GetStateOrder()
+	if err != nil {
+		return nil, err
+	}
+
+	orderB, err := b.State.GetStateOrder()
+	if err != nil {
+		return nil, err
+	}
+
+	if orderB > orderA {
```

Comment:
If this function is called "latest" shouldn't it be checking the dates?
Good example where a unit test would help clarify what is meant.

Reply:
All the tombstones start with the pending state and can only move to cancelled or processed. Since those are the only ways they can transition, we can get the newest tombstone solely based on the state. 

This function was meant to be comparing 2 tombstones with the same request ID but different states. With that function name, it can be slightly misleading and confusing. To counter this confusion, I've refactored the only function that called this function, and put the logic inside there (`GetAllTombstonesForUser`). Hopefully it is more clear. 

Side note: I also refactored `GetAllTombstonesForUser` so it only reads the tombstone contents after all the duplicate files have been filtered out. This should reduce the total number of calls to the object store. 

I know with the extra refactoring this can cause more confusion when reviewing PR's, sorry about that. 

> Good example where a unit test would help clarify what is meant.

The unit tests for the calling function were testing this functionality. I see how not having it for the individual function could cause confusion. Having them for this function could have made things more clear.